### PR TITLE
Milestone 3 - add associations and remaining classes from UML diagram

### PR DESCRIPTION
**Pull request by ram1117**
Hi, in this PR, I have implemented the following. Please review and provide feedback.

### 1.create `Classroom` class and establish `has-many`/`belongs-to` relationship with `Student` class. Adding a student to classroom will set the student's classroom. Setting classroom to the student will add the student to that classroom.

image
image

### 2. Create class `Book` and intermediate class `Rental` and establish `many-to-many` relationship between `Book` and `Person` classes using intermediate class `Rental`. Book and Person can have multiple rentals. Each Rental has one Book, one Person and a date value associated with it.

image

image

image

### CODE REVIEW

File: modules/student.rb [hunk 1]
```diff
@@ -1,9 +1,12 @@
 require_relative './person'
 
 class Student < Person
```
Comment (rloterh) [inline modules/student.rb:3]: - [ ] [OPTIONAL] It's a good practice we mark the _classroom_ object with `attr_reader`, as we only intend to _access (get)_. this object. Doing so will prevent the _classroom_ object from unnecessary modifications.

File: modules/student.rb [hunk 2]
```diff
@@ -1,9 +1,12 @@
...
+  attr_accessor :classroom
+
   def initialize(age:, classroom:, name: 'unknown', parent_permission: true)
     super(name: name, parent_permission: parent_permission, age: age)
     @classroom = classroom
+    classroom.students << self unless classroom.students.include?(self)
   end
 
   def play_hooky
```
Comment (rloterh) [inline modules/student.rb]: - [ ] _Student_ object belongs to the _Classroom_. Let's establish the belong to relationship accordingly in a standalone method like the example below:
```ruby
class Student < Person
  attr_reader :classroom
  .
  .
   def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
end
```

File: modules/book.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (rloterh) [inline modules/book.rb]: - [ ] `rental` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _modify (set)_ it. Kindly ensure to improve upon this aspect.

File: modules/rental.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class Rental
+  attr_accessor :date
+  attr_reader :person, :book
```
Comment (rloterh) [inline modules/rental.rb]: - [ ] Similarly, `book` and `person` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _modify (set)_ it. Kindly ensure to improve upon this aspect.

File: modules/person.rb [hunk 5]
```diff
@@ -1,16 +1,21 @@
 require_relative './nameable'
 class Person < Nameable
-  attr_reader :id
+  attr_reader :id, :rentals
   attr_accessor :name, :age
```
Comment (rloterh) [inline modules/person.rb]: - [ ] Again, `rental` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _modify (set)_ it. Kindly ensure to improve upon this aspect.

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Ram, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 
🏁  🌴 🚦 🚗 🏫  🏡   _🏃 

## Highlights 👍  
- Good job so far establishing associations for a book library objects! 
- ReadMe looks great!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

_Kindly check the inline comments marked **[OPTIONAL]** under the review._

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: modules/student.rb (hunk 2 above)
Comment (ram1117) [reply modules/student.rb]: Hi @rloterh , Thank you for the feedback. I need some clarification on this comment. In the first comment. you suggested adding only the getter method for classroom in student class. However if I have to implement a standalone method for classroom I have to use a setter method for classroom. Also, I have established `belongs to` relationship in the Student constructor. The UML diagram doesn't show any standalone method for classroom setter.

image

File: modules/person.rb (hunk 5 above)
Comment (ram1117) [reply modules/person.rb]: Hi @rloterh , the `Person` class has a `add_rental` stand alone method. I don't see a point using a setter method for rentals.

image

File: modules/book.rb (hunk 3 above)
Comment (ram1117) [reply modules/book.rb]: Hi @rloterh , the `Book` class has a `add_rental` stand alone method. I don't see a point using a setter method for rentals.

image

File: modules/rental.rb (hunk 4 above)
Comment (ram1117) [reply modules/rental.rb]: @rloterh , the book and person value is not going to change once a `Rental` instance is created. I don't see a point in defining setter for them. Project requirement is to add getter,setter for the date value.

image

File: modules/student.rb (hunk 2 above)
Comment (rloterh) [reply modules/student.rb]: Hi @ram1117 
Thanks for approaching me on the same. Establishing the **belong to** relationship within the student constructor isn't an invalid approach, however, consider a cleaner way of implementing that, I believe you agree we'll go by the standalone approach.

Regarding the same, I'd like you to ignore my optional remarks (classroom marked with _attr_reader_). I apologize for missing this step 🙏

Comment (ram1117) [reply modules/student.rb]: Hi @rloterh . I don't agree. classroom is a required value to instantiate a student object. The student object already has a classroom value assigned in the constructor. If i were to use a standalone method to set classroom value again, the value is prone to change again. I don't see a point why someone would assign classroom value once creating the student object and change again using the standalone method.

File: modules/book.rb (hunk 3 above)
Comment (rloterh) [reply modules/book.rb]: Hi @ram1117 
Over here, we're both reading and same time writing the `add_rental` method. You can optionally look at this [oop article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor).

**Review by mukhammadrizooff — APPROVED** [review]
Hi @ram1117, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
- No linter errors ✔️ 
- Correct GitFlow ✔️ 
- Good  implementations ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: modules/rental.rb (hunk 4 above)
Comment (rloterh) [reply modules/rental.rb]: Hi @ram1117 
Again, we're writing and then read the `book` and `person` objects here, and therefore _attr_accessor`_ is required. You can optionally look at this [oop article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor).

File: modules/person.rb (hunk 5 above)
Comment (rloterh) [reply modules/person.rb]: @ram1117 
Like informed above, we writing to `rental` and the access the rental, hence we need attr_accessor in this instance. You can optionally read this [oop article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor) for the same.

**Pull request merged**